### PR TITLE
Custom Backend

### PR DESCRIPTION
**Pull request by lsuyean**
## Overview

In this PR, I create a custom backend to run robotouille to improve functionality that PDDL may not be able to provide.

## Changes Made

- Added 6 new classes
   - Object
   - Predicate
   - Action
   - Special Effect
   - State
   - Domain
- Created a new Robotouille env (still hardcoded, not parsing from json, just to check if the backend works)
- Edited canvas, renderer, simulator, input, and wrapper to work with new backend
  - Need to do some extra work to ensure Robotouille still works with pddl

## Test Coverage

All environments work
- Creates and renders correct state, predicates, objects
- Able to recognize goal states and close when reached

RobotouilleSimulator2024-02-1214-50-00-ezgif com-video-to-gif-converter

## Next Steps

- Removing hard code from backend by creating json file format to create domain
- Ensure that Robotouille still works with PddlGym

### CODE REVIEW

File: pddlgym/backend/state.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        """
+        new_special_effect = special_effect.copy(obj, args)
+        if new_special_effect not in self.special_effects:
+            self.special_effects.append(new_special_effect)
+        current = self.special_effects[self.special_effects.index(new_special_effect)]
+        current.update(self, active=True)
+        if current.completed:
+                self.special_effects.remove(current)
+
+    def check_goal(self):
```
Comment (chalo2000) [inline pddlgym/backend/state.py:213]: `def is_goal_reached(self):`

File: pddlgym/backend/state.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        """
+        Checks if the state satisfies the goal.
+
+        Returns:
+            bool: True if the state satisfies the goal, False otherwise.
+        """
+        satisfied = False
+        for goal_set in self.goal:
```
Comment (chalo2000) [inline pddlgym/backend/state.py:221]: what is a goal set? say more in the docstring or add comments

File: pddlgym/backend/state.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        if new_special_effect not in self.special_effects:
+            self.special_effects.append(new_special_effect)
+        current = self.special_effects[self.special_effects.index(new_special_effect)]
+        current.update(self, active=True)
+        if current.completed:
+                self.special_effects.remove(current)
+
+    def check_goal(self):
+        """
+        Checks if the state satisfies the goal.
```
Comment (chalo2000) [inline pddlgym/backend/state.py:215]: Returns whether the goal is satisfied in the current state

File: pddlgym/backend/state.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+        Returns:
+            bool: True if the state satisfies the goal, False otherwise.
+        """
+        satisfied = False
+        for goal_set in self.goal:
+            for goal in goal_set:
+                if not self.check_predicate(goal):
+                    break
```
Comment (chalo2000) [inline pddlgym/backend/state.py:224]: no need for breaks, only returns. rewrite this function where the last return statement is `return False` and then figure out where you would put a `return True` inside without breaking.

File: pddlgym/backend/state.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        Steps the state forward by applying the effects of the action.
+
+        Args:
+            action (Action): The action to apply the effects of.
+            args (Dictionary[str, Object]): The arguments for the action.
+        
+        Returns:
+            new_state (State): The successor state.
+            done (bool): True if the goal is reached, False otherwise.
+        """
```
Comment (chalo2000) [inline pddlgym/backend/state.py:261]: Raises AssertionError

File: pddlgym/backend/state.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        for special_effect in self.special_effects:
```
Comment (chalo2000) [inline pddlgym/backend/state.py:262]: If you move this special effect code after taking an action, it takes care of the issue where cooking takes 1 more timestep than in the configuration. Cutting seems to work as originally expected as well. Test for yourself

File: robotouille/env.py [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+    true_predicates += build_identity_predicates(environment_json)
+    true_predicates += build_location_predicates(environment_json)
+    true_predicates += build_stacking_predicates(environment_json)
+    goal = build_goal(environment_json)
+
+    state = State(domain, objects, true_predicates, goal)
+
+    return state
+
+class RobotouilleEnv(gym.Env):
```
Comment (chalo2000) [inline robotouille/env.py:467]: This needs to replace RobotouilleWrapper, not be wrapped by it.

File: robotouille/env.py [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+from pddlgym.backend.special_effect import ConditionalEffect, DelayedEffect, RepetitiveEffect
+from environments.env_generator.builder import entity_to_entity_field, create_unique_and_combination_preds, create_combinations
+import copy
+
+import gym
+from gym import spaces
+import pygame
+import numpy as np
+
+STATION_FIELD = "stations"
```
Comment (chalo2000) [inline robotouille/env.py:16]: Can we move all the hardcoding to a separate file and put a TODO in there that the values will be extracted from the domain JSON once we create it? The only thing in this file should be RobotouilleEnv (you can import what you need from the other file)

File: utils/robotouille_input.py [hunk 9]
```diff
@@ -20,37 +28,46 @@ def create_action_from_control(env, obs, action, renderer):
     Returns:
         A valid action.
     """
-    if len(action) == 0: return
-    valid_actions = list(env.action_space.all_ground_literals(obs))
-    str_valid_actions = list(map(str, valid_actions))
+    if len(action) == 0: return None, None
```
Comment (chalo2000) [inline utils/robotouille_input.py:31]: why are 2 things being returned? reflect this change in the docstring

File: utils/robotouille_input.py [hunk 10]
```diff
@@ -20,37 +28,46 @@ def create_action_from_control(env, obs, action, renderer):
...
+    valid_actions = (obs.get_valid_actions())
```
Comment (chalo2000) [inline utils/robotouille_input.py:32]: remove unnecessary parentheses

File: utils/robotouille_input.py [hunk 11]
```diff
@@ -20,37 +28,46 @@ def create_action_from_control(env, obs, action, renderer):
...
     action = action[0]
+    for predicate, value in obs.predicates.items():
```
Comment (chalo2000) [inline utils/robotouille_input.py:34]: `for literal, is_true in obs.predicates.items():`

File: utils/robotouille_input.py [hunk 12]
```diff
@@ -20,37 +28,46 @@ def create_action_from_control(env, obs, action, renderer):
...
         clicked_station = renderer.canvas.layout[layout_pos[1]][layout_pos[0]]
-        loc_tuples = [str_valid_action.split(",") for str_valid_action in str_valid_actions]
-        locs = [loc_tuple[2].split(":")[0] if 'station' in loc_tuple[2] else loc_tuple[3].split(":")[0] for loc_tuple in loc_tuples]
-        if clicked_station in locs:
-            index = locs.index(clicked_station)
-            while 'cook(' in str_valid_actions[index] or 'cut(' in str_valid_actions[index] or 'fry(' in str_valid_actions[index] or 'fry_cut_item(' in str_valid_actions[index]:
-                # We look for fry( instead which should be guaranteed to only match the fry action, and not the fryer
-                index = locs.index(clicked_station, index+1)
-            return str_valid_actions[index]
+        for args in valid_actions[ACTIONS[0]]:
```
Comment (chalo2000) [inline utils/robotouille_input.py:42]: This file is very hard to understand because of indexing into `ACTIONS`. I see you're using this as the action space as well; can you create a dictionary that maps strings to the actions so it's easier to understand what is going on, e.g. `for args in valid_actions[ACTIONS_DICT['move']]`

File: pddlgym/backend/rep_sfx_test.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
+from predicate import Predicate
```
Comment (chalo2000) [inline pddlgym/backend/rep_sfx_test.py:1]: Add a docstring to the top of this file describing what this test is for

File: pddlgym/backend/pred_test.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
+from predicate import Predicate
```
Comment (chalo2000) [inline pddlgym/backend/pred_test.py:1]: Add a docstring to the top of this file describing what this test is for. Also add asserts for what is expected at the end.

File: pddlgym/backend/delayed_sfx_test.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
+from predicate import Predicate
```
Comment (chalo2000) [inline pddlgym/backend/delayed_sfx_test.py:1]: Add a docstring to the top of this file describing what this test is for

File: pddlgym/backend/action_test.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
+from predicate import Predicate
```
Comment (chalo2000) [inline pddlgym/backend/action_test.py:1]: Add a docstring to the top of this file describing what this test is for

**Review by chalo2000 — CHANGES_REQUESTED** [review]
I tested myself and all the environments work with the new backend 😄 happy we have functionally achieved this milestone! Now let's clean some stuff up and fix minor bugs. Review my comments and address these higher-level issues
1. The environments work but there are some small bugs I noticed, particularly
- Cook time does not reflect the configuration. It takes 4 timesteps after I start the cook by pressing 'e' for the patty to cook even though it should take 3.
- An issue we both overlooked is that the patty continues to cook even when removed from the stove. In the original, I intentionally cease cooking when the patty is removed from the stove. Make a Git issue for creating a special effect to remove other special effects for this edge case.

2. The custom backend should not be under the `pddlgym` directory since they are not related. Put it at the top-level of the project
- Related to this, the tests in `backend` should not be included with the other files. Make another directory inside `backend` called `testing` and put them there along with docstring explanations.

File: pddlgym/backend/action.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+            new_state (State): The successor state.
+        """
+        assert self.check_if_valid(state, args)
+            
+        for effect, value in self.immediate_effects.items():
+            pred = Predicate(effect.name, effect.types, [args[param] for param in effect.params])
+            state.update_predicate(pred, value)
+
+        for special_effect in self.special_effects:
+            obj = args[special_effect.obj]
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply pddlgym/backend/action.py:152]: When I create predicate definitions for effects, I pass in placeholder objects into the init to keep track of which effects should apply to which objects (for example move lettuce from s1 to s2, we dont yet know what s1 or s2 are but they have different effects) Args is a dictionary which looks like {s1: table1, s2:table2}. So here, I want to apply the special effect to the right argument. The special_effect.obj is something like s1, and i am accessing the actual object by looking at args, and using that object to update in the state.

File: pddlgym/backend/state.py [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+        If the special effect is not in the state, it is added.
+        The special effect is then updated as an active update.
+
+        Args:
+            special_effect (SpecialEffect): The special effect to update.
+            obj (Object): The object to update the special effect for.
+            args (Dictionary[Object, Object]): The arguments for the special
+                effect.
+        """
+        new_special_effect = special_effect.copy(obj, args)
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply pddlgym/backend/state.py:205]: Similar to the explanation about filler objects, I need to replace the filler objects in the special effects with the actual objects being applied on. I will rename copy() to be clearer

File: utils/robotouille_input.py [hunk 19]
```diff
@@ -18,39 +19,51 @@ def create_action_from_control(env, obs, action, renderer):
...
         clicked_station = renderer.canvas.layout[layout_pos[1]][layout_pos[0]]
-        loc_tuples = [str_valid_action.split(",") for str_valid_action in str_valid_actions]
-        locs = [loc_tuple[2].split(":")[0] if 'station' in loc_tuple[2] else loc_tuple[3].split(":")[0] for loc_tuple in loc_tuples]
-        if clicked_station in locs:
-            index = locs.index(clicked_station)
-            while 'cook(' in str_valid_actions[index] or 'cut(' in str_valid_actions[index] or 'fry(' in str_valid_actions[index] or 'fry_cut_item(' in str_valid_actions[index]:
-                # We look for fry( instead which should be guaranteed to only match the fry action, and not the fryer
-                index = locs.index(clicked_station, index+1)
-            return str_valid_actions[index]
+        for args in valid_actions[action_dict['move']]:
```
Comment (chalo2000) [inline utils/robotouille_input.py:38]: much more readable! however, the repetitive for loops to check `valid_actions` makes this function look more complicated than it needs to be. usually a signal that there is a simpler way to do this. can you think of a simpler way to do this to cut down on all the for loops?

**Review by chalo2000 — CHANGES_REQUESTED** [review]
Looking much better! Thank you for making the issues - can you include them in this PR so we have a reference to when they were created? Requesting changes in hopes that we can remove the sys hacks and to see a crisper structure for `robotouille_input`

File: backend/domain.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    def __init__(self):
+        """
+        Initializes a domain object with default values.
+        """
+        self.name = ""
+        self.object_types = []
+        self.predicates = []
+        self.actions = []
+
+    def are_valid_object_types(self, object_types):
```
Comment (chalo2000) [inline backend/domain.py:19]: This should also return a boolean given the name of the function suggests so

File: backend/special_effect.py [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+        """
+        if active: return
+        for condition, value in self.condition.items():
+            if state.predicates[condition] != value:
+                return
+        for effect, value in self.effects.items():
+            state.update_predicate(effect, value)
+        self.completed = True
+
+class CreationEffect(SpecialEffect):
```
Comment (chalo2000) [inline backend/special_effect.py:352]: when we implement this in the future let's also make separate files for each special effect. file size will get enormous if we don't 😃

File: robotuille_domain.json [hunk 22]
```diff
@@ -0,0 +1,573 @@
+{
```
Comment (chalo2000) [inline robotuille_domain.json:1]: remove this from this PR, it does not belong here

**Review by chalo2000 — APPROVED** [review]
resolve comments and then LGTM

File: backend/domain.py (hunk 20 above)
Comment (lsuyean) [reply backend/domain.py:19]: Changed accordingly

File: robotuille_domain.json (hunk 22 above)
Comment (lsuyean) [reply robotuille_domain.json:1]: removed

File: backend/special_effect.py (hunk 21 above)
Comment (lsuyean) [reply backend/special_effect.py]: I made new files for each subclass and placed the subclasses in a folder backend.special_effects
